### ISSUE TITLE
HTTP/2: infinite retry with 2 parallel requests on one connection receiving GOAWAY with ENHANCE_YOUR_CALM error code

### ISSUE THREAD

**Issue by selivni** [curl/curl#5119]
### I did this

I am sending two parallel requests with huge URLs (about 5k symbols) to a remote Nginx server by one connection, exceeding its http2_max_field_size (4kb by default).

### I expected the following

I expect curl to return an error after receiving a `GOAWAY` from server.

### What seems to actually happen

1. Curl sends two parallel requests to Nginx
2. Nginx fails to process **first** request and returns `GOAWAY` with `ENHANCE_YOUR_CALM` error code.
3. Nghttp2 seems to ignore the error code and only calls on_stream_close() with `REFUSED_STREAM` error code for stream related to the **second** request (which is partially correct since Nginx didn't try to process it).
4. After this, curl receives TLS close notify and decides to retry **both** requests, because the connection they run on is marked as reused.
5. Since both requests are "bad" for Nginx, curl will keep failing to finish them even if it tries to change the order the requests are processed.

### Versions

libcurl 7.65.2
Nginx 1.10.3, 1.12.*, 1.14.*

### Logs

libcurl logs (with `DEBUG_HTTP2` and `CURLOPT_VERBOSE`)
https://gist.github.com/selivni/880049e63a23f2fe7a69536c06e62c86

### How to reproduce

I reproduced the issue using [this piece of c++ code](https://gist.github.com/selivni/a22930da545bc0ce883e16286b573891), [this Nginx config](https://gist.github.com/selivni/32e171651fdc254786af4a90c59205d2), and [this URL](https://gist.githubusercontent.com/selivni/f5ab044f0911b599ad29a30d75fb5413/raw/fe63bc6bbb88b0d8bd395de212f81a2a974a672c/curl_http2_bad_url.txt).

If you change `NUM_REQUESTS` value to 1, everything will work fine and curl will successfully fail the request with "Empty reply from server" error.

### Operating system

I've been able to reproduce this on Android and Darwin (Catalina 10.15.3) with Nginx running on Ubuntu 16.04.6.

**Comment by jay** [comment]
This may be a dupe of the bug fixed in #5074. Can you try master?

**Comment by selivni** [comment]
> This may be a dupe of the bug fixed in #5074. Can you try master?

I'll try master, but it will take some time.
Meanwhile I applied the patch to my version of curl and it didn't work: after two retries `conn` seems to get reinitialized and the cycle continues.

`
conn address: 0x8c835a00

conn address: 0x8c835a00

conn address: 0x8c000000
conn address: 0x8c000000
conn address: 0x8d800400
conn address: 0x8d800400
conn address: 0x8f00ae00
conn address: 0x8f00ae00
conn address: 0x8f00f000
conn address: 0x8f00f000
conn address: 0x8d807600
conn address: 0x8d807600
conn address: 0x8c83a600
conn address: 0x8c83a600
conn address: 0x8f004000
conn address: 0x8f004000
conn address: 0x8e004c00
conn address: 0x8e004c00
conn address: 0x8f005800
conn address: 0x8f005800
conn address: 0x8c806800
conn address: 0x8c806800
conn address: 0x8f005800
conn address: 0x8f005800
conn address: 0x8c806800
conn address: 0x8c806800
conn address: 0x8d821800
conn address: 0x8d821800
`

**Comment by bagder** [comment]
What's that output (from) ?

**Comment by selivni** [comment]
> What's that output (from) ?

Sorry for not specifying, I inserted some fprintf's into curl_retry_request under `if (retry)` to see what happens to `conn` and `conn->retrycount` and then grep'd the output. It seems that each connection only reaches 2 retries after reinitializing.

I also tried this on master and got the same behaviour.

**Comment by bagder** [comment]
So you limit curl to only use one connection per host, and you do multiple parallel requests to this same host, right?

It seems one request succeeds but what happens to the next? The log is a bit hard to follow. Why and when is the stream rejected due to REFUSED_STREAM ?

**Comment by selivni** [comment]
> So you limit curl to only use one connection per host, and you do multiple parallel requests to this same host, right?

Yes, there is one connection per host with 2 parallel requests to the same host (and even the same URL).

> It seems one request succeeds but what happens to the next? The log is a bit hard to follow. Why and when is the stream rejected due to REFUSED_STREAM ?

Unfortunately, both requests are doomed since they are too big. Nginx fails to process the first one and sends `GOAWAY` with `ENHANCE_YOUR_CALM` error code and stream_id `1` (associated with the first request). This should mean that we are supposed to count first request as failed, but retry the second one as Nginx didn't get to try to process it (at least that's what [RFC](https://tools.ietf.org/html/rfc7540#section-8.1.4) says).

If I understood it correctly, nghttp2 tries to achieve this by calling `on_stream_close()` for stream_id `3` (associated with the second request) with error_code `REFUSED_STREAM` so that we know that it's safe to retry.

I have no idea why nghttp2 doesn't tell us that the first stream caught an `ENHANCE_YOUR_CALM`. It seems like curl doesn't get that the first stream is already dead until the connection gets terminated and the retry logic begins.

**Comment by selivni** [comment]
I also want to try to fix the issue, but I'm a bit confused by an amount of possible solutions. While making nghttp2 tell us about the first stream failure sounds like the right option, it could take a while until new nghttp2 version reaches curl. Would you like me to try something similar to CONN_MAX_RETRIES solution but for requests?

**Comment by bagder** [comment]
> Would you like me to try something similar to CONN_MAX_RETRIES solution but for requests?

Does it work to set `CURLOPT_MAXREDIRS` to achieve that?

**Comment by selivni** [comment]
> > Would you like me to try something similar to CONN_MAX_RETRIES solution but for requests?
> 
> Does it work to set `CURLOPT_MAXREDIRS` to achieve that?

No, looks like it doesn't work. I tried different values including `0L` and they don't seem to stop this infinite loop.

**Comment by bagder** [comment]
Then maybe figuring out why that doesn't work is the first step? Which logic decides that the transfer should be retried? I figured it would be either `Curl_retry_request` or `Curl_follow`...

**Comment by selivni** [comment]
> Then maybe figuring out why that doesn't work is the first step? Which logic decides that the transfer should be retried? I figured it would be either `Curl_retry_request` or `Curl_follow`...

It seems like the reason is in the following `if` statement in Curl_retry_request:
```
  if((data->req.bytecount + data->req.headerbytecount == 0) &&
      conn->bits.reuse &&
      (!data->set.opt_no_body
        || (conn->handler->protocol & PROTO_FAMILY_HTTP)) &&
      (data->set.rtspreq != RTSPREQ_RECEIVE))
    /* We got no data, we attempted to re-use a connection. For HTTP this
       can be a retry so we try again regardless if we expected a body.
       For other protocols we only try again only if we expected a body.

       This might happen if the connection was left alive when we were
       done using it before, but that was closed when we wanted to read from
       it again. Bad luck. Retry the same request on a fresh connect! */
    retry = TRUE;
```

In case we send only one "bad" request at a time, the `reuse` bit is not set, so the request doesn't get retried and we receive an error.
In case we send both requests at the same time, `reuse` bit **is** set, and since all the other conditions are also met in both cases, now both requests get retried (in another order, which unfortunately doesn't matter since neither of the requests can get processed).

**Comment by bagder** [comment]
Thinking about this further, if this really is an `ENHANCE_YOUR_CALM` stream error then I don't think we should do an immediate retry as that's not likely to work immediately since the error is used in situations where the server thinks the client/load is too intense. But it would still be good to be able to give some sort of signal back to the application about what happened and for the curl tool for example to consider such an error as "transient" so that it can retry again in a short while.

So a first step would be to make sure that we actually get that error from nghttp2.

**Comment by selivni** [comment]
> Thinking about this further, if this really is an `ENHANCE_YOUR_CALM` stream error then I don't think we should do an immediate retry as that's not likely to work immediately since the error is used in situations where the server thinks the client/load is too intense. But it would still be good to be able to give some sort of signal back to the application about what happened and for the curl tool for example to consider such an error as "transient" so that it can retry again in a short while.
> 
> So a first step would be to make sure that we actually get that error from nghttp2.

I can try to raise this issue in nghttp2 github. Also, what could be implemented right now is the following:

If we receive `REFUSED_STREAM` on some streams and connection gets closed, do not retry requests that didn't receive this error and have lower `stream_id` than those that did. I think it doesn't break HTTP/2 protocol. What do you think about that?

**Comment by bagder** [comment]
It sounds reasonable. But if you get refused_stream on only a few of the streams (meaning there are one or more streams  that didn't get any error), why does the connection get closed in the first place?

**Comment by selivni** [comment]
> It sounds reasonable. But if you get refused_stream on only a few of the streams (meaning there are one or more streams that didn't get any error), why does the connection get closed in the first place?

It could happen for two reasons:
1) the server has met a serious error condition (or a shutdown), making it unable to process this and any following requests. The server sends `GOAWAY` with `stream_id`, then closes the connection.
2) we encountered a network issue.

In both cases, since we received `REFUSED_STREAM`, we know that the server is having an issue, and since the requests in question have lower `stream_id` than the ones that did receive `REFUSED_STREAM`, they cannot be safely retried because they might have been already processed once (for instance, it could be something dangerous like payment processing request or just a POST request), or they may have caused this unrecoverable error. The ones that did receive `REFUSED_STREAM`, however, are definitely safe to retry since the server guarantees that it didn't start to process them.

**Comment by jay** [comment]
What happened with this? @tatsuhiro-t any comments?

**Comment by tatsuhiro-t** [comment]
I don't fully understand this issue, but the protocol wise, if you get REFUSED_STREAM error (I think nghttp2 will pass it to stream close callback), it means that client can safely resend the request.  Whether actually resending it is client decision.  Using ENHANCE_YOUR_CLAM is entirely server's discretion, and it can do anything about it.  I don't know what nginx does there.

**Comment by bagder** [comment]
@selivni does this bug still persist in current master?

**Comment by bagder** [comment]
Added to KNOWN_BUGS now
